File: dp/c/double_checked_locking.c
// Double-Checked Locking Pattern in C
// Thread-safe lazy initialization with minimal locking overhead

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <stdatomic.h>

typedef struct {
    char db_url[128];
    int max_connections;
} Config;

static atomic_intptr_t g_config = ATOMIC_VAR_INIT(0);
static pthread_mutex_t g_mutex = PTHREAD_MUTEX_INITIALIZER;

Config* get_config(void) {
    Config* cfg = (Config*)atomic_load(&g_config);

    // First check (no lock)
    if (!cfg) {
        pthread_mutex_lock(&g_mutex);

        // Second check (with lock)
        cfg = (Config*)atomic_load(&g_config);
        if (!cfg) {
            printf("[Config] Initializing (expensive)...\n");
            cfg = malloc(sizeof(Config));
            snprintf(cfg->db_url, sizeof(cfg->db_url), "postgres://localhost/mydb");
            cfg->max_connections = 10;
            atomic_store(&g_config, (intptr_t)cfg);
        }

        pthread_mutex_unlock(&g_mutex);
    }

    return cfg;
}

// pthread_once alternative (recommended in practice)
static pthread_once_t once_ctrl = PTHREAD_ONCE_INIT;
static Config* once_config = NULL;

void init_once_config(void) {
    printf("[Once] Initializing config...\n");
    once_config = malloc(sizeof(Config));
    snprintf(once_config->db_url, sizeof(once_config->db_url), "postgres://localhost/db2");
    once_config->max_connections = 20;
}

Config* get_config_once(void) {
    pthread_once(&once_ctrl, init_once_config);
    return once_config;
}

int main(void) {
    printf("=== Double-Checked Locking ===\n");
    Config* c1 = get_config();
    Config* c2 = get_config(); // Not initialized again
    printf("URL: %s\n", c1->db_url);
    printf("Same instance: %s\n", c1 == c2 ? "yes" : "no");

    printf("\n=== pthread_once ===\n");
    Config* c3 = get_config_once();
    Config* c4 = get_config_once();
    printf("URL: %s\n", c3->db_url);
    printf("Same instance: %s\n", c3 == c4 ? "yes" : "no");

    free(c1);
    free(once_config);
    return 0;
}

File: dp/rust/src/double_checked_locking.rs
// Double-Checked Locking Pattern in Rust
// In Rust, std::sync::OnceLock (1.70+) / Once / Lazy provide this idiom safely

use std::sync::{Mutex, OnceLock};

// Expensive resource
struct Config {
    db_url: String,
    max_connections: u32,
}

impl Config {
    fn load() -> Self {
        println!("[Config] Loading (expensive operation)...");
        Config {
            db_url: "postgres://localhost/mydb".to_string(),
            max_connections: 10,
        }
    }
}

// OnceLock: the idiomatic Rust approach (no manual double-check needed)
fn global_config() -> &'static Config {
    static CONFIG: OnceLock<Config> = OnceLock::new();
    CONFIG.get_or_init(Config::load)
}

// Manual double-checked locking (educational — prefer OnceLock in practice)
struct LazyInit<T> {
    value: OnceLock<T>,
    _init_lock: Mutex<()>,
}

impl<T> LazyInit<T> {
    const fn new() -> Self {
        LazyInit {
            value: OnceLock::new(),
            _init_lock: Mutex::new(()),
        }
    }

    fn get_or_init<F>(&self, f: F) -> &T
    where
        F: FnOnce() -> T,
    {
        // First check (no lock)
        if let Some(val) = self.value.get() {
            return val;
        }
        // Acquire lock
        let _guard = self._init_lock.lock().unwrap();
        // Second check (with lock)
        self.value.get_or_init(f)
    }
}

static LAZY_VALUE: LazyInit<String> = LazyInit::new();

fn main() {
    // OnceLock approach
    println!("=== OnceLock ===");
    let cfg1 = global_config();
    let cfg2 = global_config(); // Not loaded again
    println!("DB: {}", cfg1.db_url);
    println!("Same instance: {}", std::ptr::eq(cfg1, cfg2));

    // Manual approach
    println!("\n=== Manual LazyInit ===");
    let v1 = LAZY_VALUE.get_or_init(|| {
        println!("[LazyInit] Initializing...");
        "initialized".to_string()
    });
    let v2 = LAZY_VALUE.get_or_init(|| {
        println!("[LazyInit] Should NOT print");
        "again".to_string()
    });
    println!("Value: {}", v1);
    println!("Same: {}", std::ptr::eq(v1, v2));
}
